internal/datastore/memdb/memdb.go

```diff
@@ -348,93 +34,199 @@ func NewMemdbDatastore(
-		return nil, fmt.Errorf(errUnableToInstantiateTuplestore, err)
+		return nil, err
 	}
 
-	txn.Commit()
-
 	if watchBufferLength == 0 {
 		watchBufferLength = defaultWatchBufferLength
 	}
 
 	uniqueID := uuid.NewString()
 
+	negativeGCWindow := decimal.NewFromInt(gcWindow.Nanoseconds()).Mul(decimal.NewFromInt(-1))
+
 	return &memdbDatastore{
-		db:                   db,
-		watchBufferLength:    watchBufferLength,
-		revisionQuantization: revisionQuantization,
+		db: db,
+		revisions: []snapshot{
+			{
+				revision: decimal.Zero,
+				db:       db,
+			},
+		},
 
-		gcWindowInverted: -1 * gcWindow,
-		simulatedLatency: simulatedLatency,
-		uniqueID:         uniqueID,
+		negativeGCWindow:   negativeGCWindow,
+		quantizationPeriod: decimal.NewFromInt(revisionQuantization.Nanoseconds()),
+		watchBufferLength:  watchBufferLength,
+		uniqueID:           uniqueID,
 	}, nil
 }
 
-func (mds *memdbDatastore) IsReady(ctx context.Context) (bool, error) {
-	return true, nil
-}
+type memdbDatastore struct {
+	sync.RWMutex
 
-func (mds *memdbDatastore) NamespaceCacheKey(namespaceName string, revision datastore.Revision) (string, error) {
-	return fmt.Sprintf("%s@%s", namespaceName, revision), nil
+	db             *memdb.MemDB
+	revisions      []snapshot
+	activeWriteTxn *memdb.Txn
+
+	negativeGCWindow   datastore.Revision
+	quantizationPeriod datastore.Revision
+	watchBufferLength  uint16
+	uniqueID           string
 }
 
-func (mds *memdbDatastore) Close() error {
-	mds.Lock()
-	mds.db = nil
-	mds.Unlock()
-	return nil
+type snapshot struct {
+	revision datastore.Revision
+	db       *memdb.MemDB
 }
 
-func createNewTransaction(txn *memdb.Txn) (uint64, error) {
-	var newTransactionID uint64 = 1
+func (mdb *memdbDatastore) SnapshotReader(revision datastore.Revision) datastore.Reader {
+	mdb.RLock()
+	defer mdb.RUnlock()
 
-	lastChangeRaw, err := txn.Last(tableTransaction, indexID)
-	if err != nil {
-		return 0, err
+	if err := mdb.checkRevisionLocal(revision); err != nil {
+		return &memdbReader{nil, nil, datastore.NoRevision, err}
 	}
 
-	if lastChangeRaw != nil {
-		newTransactionID = lastChangeRaw.(*transaction).id + 1
+	revIndex := sort.Search(len(mdb.revisions), func(i int) bool {
+		return mdb.revisions[i].revision.GreaterThanOrEqual(revision)
+	})
+
+	if revIndex == len(mdb.revisions) {
+		revIndex = len(mdb.revisions) - 1
 	}
 
-	newChangelogEntry := &transaction{
-		id:        newTransactionID,
-		timestamp: uint64(time.Now().UnixNano()),
+	snapshotRevision := mdb.revisions[revIndex].revision
+	roTxn := mdb.revisions[revIndex].db.Txn(false)
+
+	txSrc := func() (*memdb.Txn, error) {
+		return roTxn, nil
 	}
 
-	if err := txn.Insert(tableTransaction, newChangelogEntry); err != nil {
-		return 0, err
+	return &memdbReader{noopTryLocker{}, txSrc, snapshotRevision, nil}
+}
+
+func (mdb *memdbDatastore) ReadWriteTx(
+	ctx context.Context,
+	f datastore.TxUserFunc,
+) (datastore.Revision, error) {
+	for i := 0; i < 10; i++ {
+		var tx *memdb.Txn
```

Comment:
This variable seems unnecessary if it's available the same thing as `mdb.activeWriteTxn`. The locking semantics will be more clear if you always refer to this field instead of the local var.

Reply:
It's actually subtly different from that. Having the local var lets us distinguish between whether we allocated the transaction or someone else did in the error cases. On line 152 we wouldn't know if we had to abort the `tx`.

Reply:
It's actually subtly different from that. Having the local var lets us distinguish between whether we allocated the transaction or someone else did in the error cases. On line 152 we wouldn't know if we had to abort the `tx`.